clients/drcachesim/tests/drcachesim_unit_tests.cpp

```diff
@@ -306,9 +308,201 @@ LLC {
+                     int step_count_a, int step_size_b, int step_count_b,
+                     int loop_count = 1)
+{
+    int access_count = 0;
+    for (int i = 0; i < loop_count; ++i) {
+        memref_t ref;
+        ref.data.type = TRACE_TYPE_READ;
+        ref.data.size = 4;
+        for (int step_a = 0; step_a < step_count_a; ++step_a) {
+            for (int step_b = 0; step_b < step_count_b; ++step_b) {
+                addr_t addr = start_address + step_a * step_size_a + step_b * step_size_b;
+                ref.data.addr = addr;
+                cache.request(ref);
+                ++access_count;
+            }
+        }
+    }
+    return access_count;
+}
+
+// Convenience wrapper for a linear access pattern.
+static int
+generate_1D_accesses(cache_t &cache, addr_t start_address, int step_size, int step_count,
+                     int loop_count = 1)
+{
+    return generate_2D_accesses(cache, start_address, step_size, step_count, 0, 1,
+                                loop_count);
+}
+
+// Helper code to grab a snapshot of cache stats.
+struct CStats {
+    int_least64_t hits;
+    int_least64_t misses;
+    int_least64_t child_hits;
+};
+
+static CStats
+get_cache_stats(caching_device_stats_t &stats)
+{
+    CStats cs;
+    cs.hits = stats.get_metric(metric_name_t::HITS);
+    cs.misses = stats.get_metric(metric_name_t::MISSES);
+    cs.child_hits = stats.get_metric(metric_name_t::CHILD_HITS);
+    return cs;
+}
+
+// Creates and tests LRU caches in a range of assocativities, verifying
+// the associativity works as expected.
+void
+unit_test_cache_associativity()
+{
+    // Range of associativities to be tested.
+    constexpr int MIN_ASSOC = 1;
+    constexpr int MAX_ASSOC = 16;
+
+    constexpr int LINE_SIZE = 32;
+    constexpr int BLOCKS_PER_WAY = 16;
+
+    // Only power-of-two associativities are currently supported.
+    for (int assoc = MIN_ASSOC; assoc <= MAX_ASSOC; assoc *= 2) {
+        int total_size = LINE_SIZE * BLOCKS_PER_WAY * assoc;
+        // Test access patterns that stress increasing associativity.
+        for (uint32_t test_assoc = 1; test_assoc <= 2 * assoc; ++test_assoc) {
+            cache_lru_t cache;
+            caching_device_stats_t stats(/*miss_file=*/"", LINE_SIZE);
+            assert(cache.init(assoc, LINE_SIZE, total_size, /*parent=*/nullptr, &stats));
+            // Test start address is arbitrary.
+            addr_t start_address = test_assoc * total_size;
+
+            constexpr int NUM_LOOPS = 3; // Anything >=2 should work.
+            auto read_count =
+                generate_2D_accesses(cache, start_address, LINE_SIZE, BLOCKS_PER_WAY,
+                                     total_size, test_assoc, NUM_LOOPS);
+            CStats c_stats = get_cache_stats(stats);
+            assert(read_count == NUM_LOOPS * BLOCKS_PER_WAY * test_assoc);
+
+            int expected_misses =
+                (test_assoc <= assoc) ? BLOCKS_PER_WAY * test_assoc : read_count;
```

Comment:
(OK, because it's LRU.  Where did we request LRU --- ah ok took me a while to see the type `cache_lru_t`; kept looking for a param in the cache.init call.  Worth a comment?  Or maybe it would be obvious to a more careful reader.)

Reply:
You're right, it's worth highlighting why the miss count is so predictable.  I added a comment.